packages/stream/index.js

```diff
@@ -115,6 +115,91 @@ export const makePipe = () => {
 };
 harden(makePipe);
 
+/**
+ * @template TRead
+ * @template TWrite
+ * @template TReadReturn
+ * @template TWriteReturn
+ * @param {import('./types.js').Stream<TWrite, TRead, TWriteReturn, TReadReturn>} writer
+ * @param {import('./types.js').Stream<TRead, TWrite, TReadReturn, TWriteReturn>} reader
+ * @param {TWrite} primer
+ */
+export const pump = async (writer, reader, primer) => {
+  /** @param {Promise<IteratorResult<TRead, TReadReturn>>} promise */
+  const tick = promise =>
+    promise.then(
+      result => {
+        if (result.done) {
+          return writer.return(result.value);
+        } else {
+          // Behold: mutual recursion.
+          // eslint-disable-next-line no-use-before-define
+          return tock(writer.next(result.value));
+        }
+      },
+      (/** @type {Error} */ error) => {
+        return writer.throw(error);
+      },
+    );
+  /** @param {Promise<IteratorResult<TWrite, TWriteReturn>>} promise */
+  const tock = promise =>
+    promise.then(
+      result => {
+        if (result.done) {
+          return reader.return(result.value);
+        } else {
+          return tick(reader.next(result.value));
+        }
+      },
+      (/** @type {Error} */ error) => {
+        return reader.throw(error);
+      },
+    );
+  await tick(reader.next(primer));
+  return undefined;
```

Comment:
Why not return the value of `tick` ?

I'm curious as to what the return type actually is of tick & tock. Seems like it would be `Promise<IteratorResult<never, TReadReturn | TWriteReturn>>` depending on which stream closed first?

Reply:
That is almost certainly the exact type and I think the only useful part of that is the possibility that it terminates on an error. It seems less regrettable for now to return `Promise<undefined>` and consider the union type only if and when we find a motivating use.

Reply:
Just verifying: With the code as is, the error case is still handled correctly because the `await` will turn a rejection into a thrown error, and the `async` will turn the thrown error back into a rejection.

Reply:
Agreed, we can always add later